### DISCUSSION TITLE
Possible end-to-end architecture?

### DISCUSSION THREAD

**Discussion by sellorm** [thisisnic/shinyarrowpi#5] [Ideas]
**This is all just an idea. I may have totally overthought it!**

There are a few things we want to get into the presentation:

* Shiny in production (obvs!)
* Arrow
* R4Pi

Since we're converging on the idea of a sensor project, perhaps we should frame it as an IoT project.

```mermaid
  graph LR;
      A(Sensor Bundle) --sensor changes-->B(Data ingestion API);
      A --regular heartbeat--> C(Heartbeat API);
      A --Daily full data in arrow--> E("File storage (feather/parquet)")
      B-->D[(Database)];
      C-->D;
      E --> F(Create a model?)
      D --> G(Production Shiny app)
      F --> G
      A --> H(Local monitoring Shiny app)
```

If we frame it as an IoT project, we could suggest that the idea would be to have 100s (or more) of the sensor bundles.

## Sensor bundle (Mark)

* Raspberry Pi with the Enviro+ HAT and R4Pi.
* All sensor data is logged at a 1 minute resolution.
* Changes (and only changes) are pushed to the data ingestion API
    - This reduces bandwidth consumption in bandwidth constrained environments by not pushing redundant data when there have been no changes
* Regular heartbeats are pushed to the heartbeat API
    - This allows us to ensure that the sensor package is still alive even if no sensor data changes have been pushed lately
* Once per day (or some other interval) the fulls day's data is saved in an arrow format and pushed to cloud storage

## Local monitoring Shiny app (Mark/Nic)

Small shiny app that just shows the current state of the sensors and perhaps a small amount of historical data.

## Data ingestion API (Mark)

Takes the changed data and dumps it into a database

## Heartbeat API (Mark)

Takes the regular heartbeat data and dumps it into a database

## Create a model? (Nic)

* Do we need one?
* Can we do interesting predictions or something?
* Would this help us integrate some arrow?

## Production Shiny app (Nic)

* Live sensor data display
    - We could simulate having multiple sensor bundles
* historical data
* Incorporate model data

**Comment by thisisnic** [comment]
Is the database needed? How about we eliminate it and just store everything in Arrow?

**Reply by sellorm** [reply]
Not required at all. You might need to help me understand how we'd replace it with arrow though. I don't think I fully understand the best way to use it for something like this.

**Reply by thisisnic** [reply]
So I'm thinking literally just save data in files, with the reasoning being that for many users this is simpler than setting up a DB and with Arrow can be queried just as easily as one anyway. I probably want to nudge some of the more experienced Arrow R lot on thinking up great use case ideas here.

**Comment by thisisnic** [comment]
Had a chat with some colleagues who mentioned that the new architecture looks great + partitioning the data by date is going to be the way to go to get some really nice Arrow benefits obvious on S3

**Comment by sellorm** [comment]
Updated architecture looks something like this...

```mermaid
  graph LR;
      A(Sensor Bundle) --all data--> E("s3 (feather or parquet)")
      E --> F(Create a model?)
      F --> G
      E --> G(Production Shiny app)
      A --> H(Local monitoring Shiny app)
```

**Reply by thisisnic** [reply]
Awesome!
